lib/promoted/ruby/client/options.rb

```diff
@@ -0,0 +1,222 @@
+        #based on limit.
+        def insertion
+          @insertion
+        end
+
+        def log_user_id
+          return @log_user_id if @log_user_id
+          @log_user_id   = request.dig(:user_info, :log_user_id)
+          @log_user_id ||= request.dig('user_info', 'log_user_id')
+          @log_user_id
+        end
+
+        #A way to turn off logging.  Defaults to true.
+        def enabled?
+          @enabled
+        end
+
+         #Performs extra dev checks.  Safer but slower.  Defaults to true.
+        def perform_checks?
+          @perform_checks
+        end
+
+
+        # Default values to use on DeliveryRequests.
+        def default_request_values
+          @default_request_values
+        end
+
+        # Required as a dependency so clients can load reduce dependency on multiple
+        # uuid libraries.
+        def uuid
+          @uuid
+        end
+
+        # Defaults to 250ms
+        def delivery_timeout_millis
+          @delivery_timeout_millis
+        end
+
+        # Defaults to 3000ms
+        def metrics_timeout_millis
+          @metrics_timeout_millis
+        end
+
+        # For testing.  Allows for easy mocking of the clock.
+        def now_millis
+          @now_millis
+        end
+
+        def only_Log
+          @only_Log
+        end
+
+        def full_insertion
+          @full_insertion
+        end
+
+        def user_info
+          {
+            user_id: user_id,
+            log_user_id: log_user_id
+          }
+        end
+
+        def timing
+          @timing = {
+            client_log_timestamp: client_log_timestamp,
+            event_api_timestamp: event_api_timestamp
+          }
+        end
+
+        def request_id
+          @request_id
+        end
+
+        def log_request_params
+          {
+            platform_id: platform_id,
+            user_info: user_info,
+            timing: timing,
+            request: request,
+            insertion: compact_insertions
+          }
+        end
+
+        def request_params include_insertion: true
+          @request_params = {
+            platform_id: platform_id,
+            user_info: user_info,
+            timing: timing,
+            request_id: request_id,
+            view_id: view_id,
+            session_id: session_id,
+            insertion: compact_insertions
+          }
+          @request_params.merge!({insertion: compact_insertions}) if include_insertion
+          @request_params
+        end
+
+
+        def compact_insertions
+          @compact_insertions = []
+          insertions_to_compact = full_insertion
+          if limit
+            insertions_to_compact = insertions_to_compact[0..limit-1]
+          end
+          insertions_to_compact.each_with_index do |insertion_obj, index|
+            insertion_obj = insertion_obj.transform_keys(&:to_underscore)
+            insertion_obj = insertion_obj.transform_keys(&:to_sym)
+            insertion_obj[:user_info]    = user_info
```

Comment:
It'll be safer to have an exclusion list (rather than include everything).

We'll have clients that don't want to compact the LogRequests.  If the client wants to compact the logging statements, then the most common scenario is to copy and remove the Insertion.property fields.  We expect this to be client-specific though (which is why the Typescript one is a function).